### PR TITLE
Add a max TTL for cached entries

### LINKED ISSUE #10: Option to disable cache or refresh it asynchronously
This resolver provides a cache for IP addresses and TXT records.
Depending on the TTL set for the DNS record, this can lead to a significant delay before the resolver returns an up to date DNS record.

If this is done for latency concerns, updating the cache in the background via a goroutine would make sure the cache is refreshed regularly.
If the concern is for performance, being able to set a maximum TTL of records in the cache (0 for disabling it) would be great.

### PR DESCRIPTION
**Pull request by thibmeu**
Fixes #10.

### CODE REVIEW

**Review by marten-seemann — CHANGES_REQUESTED** [review]
Thank you for the PR!

We usually use the Options pattern to configure services. To see how this looks in practice, have a look at https://github.com/libp2p/go-tcp-transport/blob/c0bf21d689afb213896072e52f225f9126d2ba67/tcp.go#L90-L103.

In this case, we should probably have a `func WithMaxTTL(time.Duration) Option` option passed to the constructor (`func NewResolver(url string, ...Option) *Resolver`).
By default, we shouldn't limit the TTL (you can use the zero value of `r.maxCacheTTL` to mean that we're not applying any limit).

File: resolver.go [hunk 1]
```diff
@@ -81,6 +102,10 @@ func (r *Resolver) LookupIPAddr(ctx context.Context, domain string) (result []ne
 		}
 	}
 
+	if r.maxCacheTTL != NoMaxCacheTTL {
```
Comment (marten-seemann) [inline resolver.go:105]: Given that we add entries to the cache even if `ttl == 0`, can we just `maxCacheTTL == 0` to mean that we're not applying any max TTL?

File: resolver.go [hunk 2]
```diff
@@ -31,16 +32,36 @@ type txtEntry struct {
 	expire time.Time
 }
 
-func NewResolver(url string) *Resolver {
+type Option func(*Resolver) error
+
+var NoMaxCacheTTL = -1
```
Comment (marten-seemann) [inline resolver.go:37]: This could be a `const`, and should not be exported. And with my comment below, we could get rid of it altogether.

Comment (thibmeu) [reply resolver.go]: This value has now been removed.

File: resolver.go [hunk 3]
```diff
@@ -31,16 +33,34 @@ type txtEntry struct {
 	expire time.Time
 }
 
-func NewResolver(url string) *Resolver {
+type Option func(*Resolver) error
+
+func WithMaxCacheTTL(maxCacheTTL uint32) Option {
```
Comment (marten-seemann) [inline resolver.go:38]: This should be a `time.Duration`.

[the remark this answers is not in the document]
Comment (marten-seemann) [reply resolver.go:38]: We don't handle 0 correctly at the moment. We insert every value into the cache. That's something we can fix, but I don't think we have to do it in this PR.

[the remark this answers is not in the document]
Comment (aschmahmann) [reply resolver.go:38]: Yep, I'm mostly saying the semantics might want to be called out. To some extent it's reasonable to assume users will figure out that a max TTL of 0 means you'll always do a lookup but might be worth specifying.

Waiting until the fix comes in to update the comment is fine too if you're concerned about bugs involving things like Windows thinking the time difference between two events is 0 when it's just small.

[the remark this answers is not in the document]
Comment (marten-seemann) [reply resolver.go]: Ah sorry, I missed that!
If I understand @aschmahmann correctly, he was suggesting was introducing a `DisableCache` option, which internally would just set the max TTL to 0. I think that would be a good idea.

[the remark this answers is not in the document]
Comment (thibmeu) [reply resolver.go]: I also added a comment on `WithMaxCacheTTL` function. It describes the normal behaviour, has well as the behaviour of `0`.

File: resolver_test.go [hunk 4]
```diff
@@ -45,7 +49,36 @@ func TestLookupIPAddr(t *testing.T) {
...
+	if !ok {
+		t.Fatal("expected cache to be populated")
+	}
+	if !sameTXT(txt, txt2) {
+		t.Fatal("expected cache to contain the same txt entries")
+	}
+}
+
+func TestLookupCache(t *testing.T) {
+	cacheTTL := time.Duration(1) * time.Second
```
Comment (marten-seemann) [inline resolver_test.go:77]:
```suggestion
	const cacheTTL = time.Second
```

File: resolver.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (aschmahmann) [reply resolver.go]: I wasn't saying that specifically just asking for docs, but that's a reasonable idea/approach @marten-seemann 😁

**Comment by marten-seemann** [comment]
There are also a few failures in the `Go Checks` workflow that need fixing.

**Comment by marten-seemann** [comment]
Thank you!

> There is an issue with macOS on 1.16, not sure if this is due to the local server or a one off issue.

@codecov again. This tool has been causing us so many problems....

**Pull request merged**